pkg/apis/deviceplugin/v1/iaadeviceplugin_webhook.go

```diff
@@ -15,79 +15,42 @@
 	return ctrl.NewWebhookManagedBy(mgr).
 		For(r).
+		WithDefaulter(&commonDevicePluginDefaulter{
+			defaultImage: "intel/intel-iaa-plugin:" + controllers.ImageMinVersion.String(),
+		}).
+		WithValidator(&commonDevicePluginValidator{
+			expectedImage:     "intel-iaa-image",
+			expectedInitImage: "intel-idxd-config-initcontainer",
+			expectedVersion:   *controllers.ImageMinVersion,
+		}).
 		Complete()
 }
 
 // +kubebuilder:webhook:path=/mutate-deviceplugin-intel-com-v1-iaadeviceplugin,mutating=true,failurePolicy=fail,groups=deviceplugin.intel.com,resources=iaadeviceplugins,verbs=create;update,versions=v1,name=miaadeviceplugin.kb.io,sideEffects=None,admissionReviewVersions=v1
-
-var _ webhook.Defaulter = &IaaDevicePlugin{}
-
-// Default implements webhook.Defaulter so a webhook will be registered for the type.
-func (r *IaaDevicePlugin) Default() {
-	iaadevicepluginlog.Info("default", "name", r.Name)
-
-	if len(r.Spec.Image) == 0 {
-		r.Spec.Image = "intel/intel-iaa-plugin:" + iaaMinVersion.String()
-	}
-}
-
 // +kubebuilder:webhook:verbs=create;update,path=/validate-deviceplugin-intel-com-v1-iaadeviceplugin,mutating=false,failurePolicy=fail,groups=deviceplugin.intel.com,resources=iaadeviceplugins,versions=v1,name=viaadeviceplugin.kb.io,sideEffects=None,admissionReviewVersions=v1
 
-var _ webhook.Validator = &IaaDevicePlugin{}
-
-// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (r *IaaDevicePlugin) ValidateCreate() (admission.Warnings, error) {
-	iaadevicepluginlog.Info("validate create", "name", r.Name)
-
-	return nil, r.validatePlugin()
-}
-
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (r *IaaDevicePlugin) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
-	iaadevicepluginlog.Info("validate update", "name", r.Name)
-
-	return nil, r.validatePlugin()
-}
-
-// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
-func (r *IaaDevicePlugin) ValidateDelete() (admission.Warnings, error) {
-	iaadevicepluginlog.Info("validate delete", "name", r.Name)
-
-	return nil, nil
-}
-
-func (r *IaaDevicePlugin) validatePlugin() error {
-	if err := validatePluginImage(r.Spec.Image, "intel-iaa-plugin", iaaMinVersion); err != nil {
+func (r *IaaDevicePlugin) validatePlugin(ref *commonDevicePluginValidator) error {
+	if err := validatePluginImage(r.Spec.Image, ref.expectedImage, &ref.expectedVersion); err != nil {
 		return err
 	}
 
 	if len(r.Spec.ProvisioningConfig) > 0 && len(r.Spec.InitImage) == 0 {
-		return errors.Errorf("ProvisioningConfig is set with no InitImage")
+		return fmt.Errorf("%w: ProvisioningConfig is set with no InitImage", errValidation)
 	}
 
 	if len(r.Spec.InitImage) > 0 {
-		return validatePluginImage(r.Spec.InitImage, "intel-idxd-config-initcontainer", iaaMinVersion)
+		return validatePluginImage(r.Spec.InitImage, ref.expectedInitImage, &ref.expectedVersion)
```

Comment:
I think these checks have unnecessary differences between plugins, and this part could also be in the common code.

If `ref.expectedInitImage` is specified,  spec should have it, otherwise not.  Right?

Reply:
closing as resolved. please reopen if you disagree

Reply:
Yes, the improvement in this PR is enough.  Further refactoring can always be done in further PRs...